### PR TITLE
Alternative voting

### PR DESCRIPTION
**Pull request by Norfolks**
## Functionality added

- **Fast track proposals** — a new FastTrack governance flow with bond-based fees, a "For"-only sandbox pre-voting period, scheduled Monday voting, flexible majority types (simple/strong), council veto, and reviewer slashing.
- **Parallel execution** — FastTrack runs alongside the existing Classic flow, so both proposal lifecycles operate in parallel on the same contract.
- **Queue for proposals** — a shared FIFO queue where up to `max_active_proposals` occupy active slots; overflow proposals park as `Queued` and are promoted as slots free up.

### CODE REVIEW

File: venear-contract/src/delegation.rs [hunk 1]
```diff
@@ -1,59 +1,96 @@
...
+        }
+        if i > 0 && entry.account_id <= entries[i - 1].account_id {
+            return Err("Entries must be sorted ascending by account_id with no duplicates");
+        }
+        sum_bps += entry.bps as u32;
+    }
+    if sum_bps > 10_000 {
+        return Err("Total bps exceeds 10000");
+    }
+    Ok(sum_bps as u16)
```
Comment (frol) [inline venear-contract/src/delegation.rs]: I strongly recommend to forbid "as" syntax. Use `.into()`/`u64::from()`/`.try_into()`/`u64::try_from()` instead. Otherwise, you can silently introduce overflow bugs.

File: venear-contract/src/delegation.rs [hunk 2]
```diff
@@ -1,59 +1,96 @@
...
+                .pooled_add_scaled(&owner.balance, entry.bps);
+            self.internal_set_account(entry.account_id.clone(), delegate);
+        }
+
+        events::emit::delegation_change(&predecessor_id, &old_delegations, &entries);
+        owner.delegations = entries;
+        self.internal_set_account(predecessor_id.clone(), owner);
+
+        let storage_after = env::storage_usage();
+        if storage_after > storage_before {
```
Comment (frol) [inline venear-contract/src/delegation.rs]: Make sure you `.flush()` all the changes for `near_sdk::store` collections as they are lazy and don't commit changes to the storage until you drop the collection (which mostly happens after the function call, when Contract struct Drop impl is calling flush). So if you need to track the storage changes, you must call `.flush()` explicitly.

File: voting-contract/src/bond.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            ),
+            "Bond can only be claimed from expired or rejected proposals"
+        );
+
+        let bond = proposal.bond_amount;
+        let proposer_id = proposal.proposer_id.clone();
+        proposal.bond_amount = NearToken::from_yoctonear(0);
+        self.internal_set_proposal(proposal);
+
+        Promise::new(proposer_id).transfer(bond);
```
Comment (frol) [inline voting-contract/src/bond.rs]: Upgrade near-sdk to the latest version and address all clippy warnings.

One of the warnings will be about this line that will require you to either explicitly call `.detach()` here or return the value as the function result (I assume that you wanted to not fail the claim_bond function if the transfer fails, but it is not 100% clear here without the `.detach()` call).

File: voting-contract/src/lib.rs [hunk 4]
```diff
@@ -12,19 +14,41 @@ use merkle_tree::{MerkleProof, MerkleTreeSnapshot};
...
 #[derive(BorshStorageKey)]
 #[near]
 enum StorageKeys {
     Proposal,
     ProposalMetadata,
     Votes,
-    ApprovedProposals,
+    PendingQueue,
+    ActiveProposals,
 }
```
Comment (frol) [inline voting-contract/src/lib.rs:56]: DO NOT REPLACE StorageKeys, this will change the prefix for the stored collection keys and your storage will be corrupted if there are any existing `ApprovedProposals` records in the storage. For even more explicit way of defining these, you can use the following syntax:

```rust
enum StorageKeys {
    Proposal = 0,
    ProposalMetadata = 1,
    Votes = 2,
    // ApprovedProposals = 3,
    PendingQueue = 4,
    ActiveProposals = 5,
}
```

File: voting-contract/src/reviewer.rs [hunk 5]
```diff
@@ -1,37 +1,79 @@
...
-            .then(
-                ext_self::ext(env::current_account_id())
-                    .with_static_gas(GAS_FOR_ON_GET_SNAPSHOT)
-                    .on_get_snapshot(env::predecessor_account_id(), proposal_id),
+        proposal.reviewer_id = Some(env::predecessor_account_id());
+        proposal.approval_time_ns = Some(env::block_timestamp().into());
+        proposal.quorum_threshold_bps = self.config.quorum_threshold_bps;
+        proposal.quorum_floor = self.config.quorum_floor;
+
+        if proposal.bond_amount.as_yoctonear() > 0 {
```
Comment (frol) [inline voting-contract/src/reviewer.rs]: I would recommend to try to avoid using `as_yoctonear()` throughout the codebase as much as possible to prevent potential mix of invalid units.

In this specific case it is trivial:

```suggestion
        if proposal.bond_amount > NearToken::ZERO {
```

In other places, you can use `checked_mul`, `checked_div` (allows arbitrary u128 integer multiplier), `checked_add`, and `checked_sub` (allows only NearToken). So type safety can help you to prevent multiplying/dividing `NearToken * NearToken` or adding/substracting `NearToken + random integer`.

File: voting-contract/src/governance.rs [hunk 6]
```diff
@@ -147,6 +171,93 @@ impl Contract {
...
+    /// Can only be called by the owner.
+    /// Requires 1 yocto NEAR.
+    #[payable]
+    pub fn set_simple_majority_threshold_bps(&mut self, simple_majority_threshold_bps: u16) {
+        assert_one_yocto();
+        self.assert_owner();
+        require!(
+            simple_majority_threshold_bps <= 10_000,
+            "Simple majority threshold must be <= 10000 bps"
+        );
```
Comment (frol) [inline voting-contract/src/governance.rs]: nit: Consider implementing `Bps` type that will enforce the values range on construction / deserialization, and implement multiplication calculations for NearToken.

File: voting-contract/src/governance.rs [hunk 7]
```diff
@@ -147,6 +171,93 @@ impl Contract {
...
+    }
+
+    /// Updates the FastTrack sandbox duration in seconds.
+    /// Can only be called by the owner.
+    /// Requires 1 yocto NEAR.
+    #[payable]
+    pub fn set_sandbox_duration(&mut self, sandbox_duration_sec: u32) {
+        assert_one_yocto();
+        self.assert_owner();
+        self.config.sandbox_duration_ns = (sandbox_duration_sec as u64 * 10u64.pow(9)).into();
```
Comment (frol) [inline voting-contract/src/governance.rs]: nit: Consider using standard Duration type: https://doc.rust-lang.org/stable/std/time/struct.Duration.html or implement your own strict type.

**Review by frol — CHANGES_REQUESTED** [review]
The changes are massive, so I did my best effort to review it, but it is nowhere close to a full security audit.

Overall, I feel that the contracts grow too much with additional modes and queues which I would prefer to move to a third contract and compose it with a more basic voting contract instead (I would even say that SputnikDAO could have been a good foundation for that - no need for fork it, just design the voting structure through contract members).

File: venear-contract/src/delegation.rs (hunk 2 above)
Comment (Norfolks) [reply venear-contract/src/delegation.rs]: So far it looks like everything works fine without `.flush()`, should it be added just in case?

**Review by frol — APPROVED** [review]
I have reviewed the code on the high-level and at least nothing jumps out to me. Yet, it is not a formal security audit.

File: venear-contract/src/delegation.rs (hunk 2 above)
Comment (frol) [reply venear-contract/src/delegation.rs]: That is suspicious. I would recommend to dig into it to understand why it works as it does not meet the expectations.

File: voting-contract/src/votes.rs [hunk 8]
```diff
@@ -88,10 +113,11 @@ impl Contract {
             )
         );
 
-        // Note, don't refund 1 yoctoNEAR if changing the vote.
         if attached_deposit > near_add(storage_added, NearToken::from_yoctonear(1)) {
             let refund = near_sub(attached_deposit, storage_added);
```
Comment (think-in-universe) [inline voting-contract/src/votes.rs]: This refund retains only `storage_added`, but the branch condition reserves `storage_added + 1 yocto`. As written, any caller who overpays gets the required 1 yocto refunded too. If the 1 yocto is intended to be retained as the access-key/security marker, this should subtract both `storage_added` and `NearToken::from_yoctonear(1)` from the refund amount.

File: voting-contract/src/votes.rs [hunk 9]
```diff
@@ -111,4 +146,62 @@ impl Contract {
     pub fn get_vote(&self, account_id: AccountId, proposal_id: ProposalId) -> Option<u8> {
         self.votes.get(&(account_id, proposal_id)).cloned()
     }
+
+    /// Fetches a fresh veNEAR snapshot for a proposal that's already in Sandbox/Voting without
+    /// a snapshot, and optionally casts a vote in the same transaction.
+    #[payable]
+    pub fn take_snapshot_and_vote(
```
Comment (think-in-universe) [inline voting-contract/src/votes.rs:160]: `take_snapshot_and_vote(..., None)` is payable, but the snapshot-only path does not use or refund `env::attached_deposit()`. A caller/frontend attaching more than intended would strand that deposit in the contract. Consider requiring zero deposit when `vote.is_none()`, or refunding the full attached deposit before returning the snapshot promise.

File: venear-contract/src/delegation.rs [hunk 10]
```diff
@@ -1,59 +1,190 @@
...
+        owner.delegations = entries;
+        self.internal_set_account(predecessor_id.clone(), owner);
 
-        account.delegation = Some(AccountDelegation {
-            account_id: receiver_id,
-        });
-        self.internal_set_account(predecessor_id, account);
+        self.tree.flush();
+        let storage_after = env::storage_usage();
+        if storage_after > storage_before {
```
Comment (think-in-universe) [inline venear-contract/src/delegation.rs]: This only refunds when storage grows. Since `set_delegations` requires a positive deposit for every call, callers who replace with the same-size delegation set or shrink/remove delegations keep the entire attached deposit locked in the contract. Please refund the full attached amount when `storage_after <= storage_before`.

**Review by think-in-universe — COMMENTED** [review]
I reviewed the payment/deposit handling paths in this PR and left a few comments where funds can be refunded incorrectly or retained unexpectedly.

**Review by aiworkbot — COMMENTED** [review]
Review comments from my PR #63 pass. I focused on feature behavior and contract-risk paths; `cargo test -p voting-contract` passed locally with 179 tests.

File: voting-contract/src/upgrade.rs [hunk 11]
```diff
@@ -55,54 +62,58 @@ struct OldContract {
...
             config: Config {
                 venear_account_id: old.config.venear_account_id,
                 reviewer_ids: old.config.reviewer_ids,
-                council_ids: COUNCIL_MEMBERS.iter().map(|s| s.parse().unwrap()).collect(),
+                council_ids: old.config.council_ids,
                 owner_account_id: old.config.owner_account_id,
-                voting_duration_ns: old.config.voting_duration_ns,
-                timelock_duration_ns: U64(TIMELOCK_DURATION_NS),
+                classic_voting_duration_ns: U64(DEFAULT_CLASSIC_VOTING_DURATION_NS),
+                fast_track_voting_duration_ns: U64(DEFAULT_FAST_TRACK_VOTING_DURATION_NS),
```
Comment (aiworkbot) [inline voting-contract/src/upgrade.rs:89]: This migration reads the old `voting_duration_ns`, `timelock_duration_ns`, and `proposal_expiration_ns`, but then replaces the live classic/timelock settings with defaults. That can silently change deployed governance timing during upgrade. I would preserve `old.config.voting_duration_ns` into `classic_voting_duration_ns`, preserve `old.config.timelock_duration_ns`, and preserve `old.config.proposal_expiration_ns` into `classic_proposal_expiration_ns`; only the newly introduced FastTrack fields need seeded defaults.

File: voting-contract/src/queue.rs [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        for &id in self.pending_queue.iter() {
+            if virtual_active_count >= self.config.max_active_proposals {
+                break;
+            }
+
+            let mut proposal = self.internal_expect_proposal_updated(id);
+
+            // Pre-existing empty slots start at `now`.
+            let start_time = freed_slot_times.pop_front().unwrap_or(now).into();
+
```
Comment (aiworkbot) [inline voting-contract/src/queue.rs:95]: Backdating queued proposals to the freed slot time means a delayed `advance_queue()` can consume a queued proposal voting/sandbox window before participants ever had a real chance to snapshot/vote. Because the proposal is immediately updated against current `now`, a proposal can become terminal on promotion if no transaction advanced the queue in time. Consider activating at `now`, bounding the backdate to a grace period, or making the keeper/liveness requirement explicit and enforced operationally.

File: voting-contract/src/votes.rs [hunk 13]
```diff
@@ -88,10 +113,14 @@ impl Contract {
...
-        // Note, don't refund 1 yoctoNEAR if changing the vote.
         if attached_deposit > near_add(storage_added, NearToken::from_yoctonear(1)) {
-            let refund = near_sub(attached_deposit, storage_added);
-            Promise::new(env::predecessor_account_id()).transfer(refund);
+            let refund = near_sub(
+                attached_deposit,
+                near_add(storage_added, NearToken::from_yoctonear(1)),
+            );
+            Promise::new(env::signer_account_id())
+                .transfer(refund)
```
Comment (aiworkbot) [inline voting-contract/src/votes.rs]: Refunding all excess deposits to `signer_account_id()` fixes the self-call path from `take_snapshot_and_vote()`, but it changes behavior for ordinary contract-to-contract calls into `vote()`: the predecessor contract supplies the attached deposit, while the signer can be a different original user. I would choose the recipient by context: refund signer only when `predecessor_account_id == current_account_id` for the internal chained vote, otherwise refund `predecessor_account_id`.

File: voting-contract/src/proposal.rs [hunk 14]
```diff
@@ -287,29 +398,38 @@ impl Contract {
...
+            sandbox_threshold_bps: Bps::ZERO,
         };
         let storage_usage = env::storage_usage();
         self.proposals.push(proposal.into());
         self.proposals.flush();
         self.proposal_metadata.push(metadata.into());
         self.proposal_metadata.flush();
-        let updated_storage_usage = env::storage_usage();
+        let updated_storage_usage = env::storage_usage() + SNAPSHOT_STORAGE_RESERVE_BYTES;
         let storage_added = updated_storage_usage.saturating_sub(storage_usage);
```
Comment (aiworkbot) [inline voting-contract/src/proposal.rs:426]: This relies on `SNAPSHOT_STORAGE_RESERVE_BYTES = 101` matching the future serialized `SnapshotAndState` write, but `on_get_snapshot()` stores the actual snapshot without measuring or charging the storage delta. The reserve may match the current layout, but it is easy for future enum/layout changes to shift that cost to the contract silently. Please add a test/assertion that the actual callback storage delta stays within the reserve for supported state versions, or charge/refund actual storage in the callback.

File: voting-contract/src/queue.rs (hunk 12 above)
Comment (Norfolks) [reply voting-contract/src/queue.rs:95]: This logic is intentional; also, advance_queue() is not necessary to be triggered all the time, as there is a virtual promotion, so voters have a full voting window.

File: voting-contract/src/proposal.rs [hunk 15]
```diff
@@ -25,27 +31,51 @@ pub enum ProposalAction {
...
+    pub quorum_threshold_bps: Bps,
+    pub quorum_floor: NearToken,
+    pub approval_threshold_bps: Bps,
+    pub actions: Option<Vec<ProposalAction>>,
+    pub flow: ProposalFlow,
+    // Classic only
+    pub timelock_duration_ns: U64,
+    // FastTrack only
+    pub sandbox_start_time_ns: Option<U64>,
+    pub bond_amount: NearToken,
```
Comment (think-in-universe) [inline voting-contract/src/proposal.rs:68]: Because `bond_amount` is reset to `0` after claimed, it would be better add comment to explain the `bond_amount` is the unclaimed NEAR amount.

File: voting-contract/src/proposal.rs [hunk 16]
```diff
@@ -171,12 +144,84 @@ impl VoteStats {
     }
 }
 
+/// Returns the next voting start time. In sandbox test mode, starts after 120 seconds.
+/// Otherwise, starts on the next Monday 00:00 CET (fixed UTC+1) strictly after `after_ns`.
```
Comment (think-in-universe) [inline voting-contract/src/proposal.rs:149]: Why UTC+1 timezone instead of UTC? CET may move to UTC+2 in summer?

File: voting-contract/src/governance.rs [hunk 17]
```diff
@@ -95,32 +108,40 @@ impl Contract {
...
+        self.config.timelock_duration_ns = (u64::from(timelock_duration_sec) * 10u64.pow(9)).into();
     }
 
-    /// Updates the proposal expiration duration in seconds.
+    /// Updates the Classic proposal expiration duration in seconds.
     /// Set to 0 to disable proposal expiration.
     /// Can only be called by the owner.
     /// Requires 1 yocto NEAR.
     #[payable]
     pub fn set_proposal_expiration(&mut self, proposal_expiration_sec: u32) {
```
Comment (think-in-universe) [inline voting-contract/src/governance.rs]: Should we make the function name more explicit that it's for Classical proposal expiration?

File: voting-contract/src/upgrade.rs [hunk 18]
```diff
@@ -1,47 +1,51 @@
...
 
-const DEFAULT_QUORUM_THRESHOLD_BPS: u16 = 3500;
-const DEFAULT_QUORUM_FLOOR_NEAR: u128 = 1000;
-const DEFAULT_APPROVAL_THRESHOLD_BPS: u16 = 5000;
-const TIMELOCK_DURATION_NS: u64 = 14 * 24 * 60 * 60 * 1_000_000_000; // 14 days
-const PROPOSAL_EXPIRATION_NS: u64 = 7 * 24 * 60 * 60 * 1_000_000_000; // 7 days
+// Defaults applied when migrating a contract that predates the merged flow.
+const DEFAULT_BOND_AMOUNT_NEAR: u128 = 100;
+// TODO: change the migration default treasury once the real treasury account is set up.
+const DEFAULT_TREASURY_ACCOUNT_ID: &str = "norfolks.near";
```
Comment (think-in-universe) [inline voting-contract/src/upgrade.rs:19]: Is it intentional to use this personal account as the default treasury account ID?

**Review by think-in-universe — COMMENTED** [review]
LGTM with some minor issues.

File: voting-contract/src/proposal.rs (hunk 16 above)
Comment (Norfolks) [reply voting-contract/src/proposal.rs:149]: It was just input requirement
This one is fixed to +1

File: voting-contract/src/upgrade.rs (hunk 18 above)
Comment (Norfolks) [reply voting-contract/src/upgrade.rs:19]: It will be changed before merging to main, as soon as we have treasury account address.

**Review by think-in-universe — APPROVED** [review]
